### PR TITLE
system parameters: Change defaults to be closer to LaunchDarkly settings in production

### PR DESCRIPTION
**Pull request by def-**
to make self-hosted Materialize and the Materialize Emulator more useful without having to set many system parameters.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/adapter-types/src/dyncfgs.rs
Comment (def-) [inline src/adapter-types/src/dyncfgs.rs:69]: This might be expensive, but we are also doing it in production.

File: src/persist-client/src/cli/admin.rs
Comment (def-) [inline src/persist-client/src/cli/admin.rs:674]: This might also make us slower, but set in production.

File: src/persist-client/src/internal/cache.rs
Comment (def-) [inline src/persist-client/src/internal/cache.rs:35]: Increases our memory usage but improves performance.

File: src/persist-client/src/stats.rs
Comment (def-) [inline src/persist-client/src/stats.rs:32]: Slower, but I consider the chance of noticing correctness issues to be more important.

File: src/sql/src/session/vars/definitions.rs
Comment (def-) [inline src/sql/src/session/vars/definitions.rs:1394]: This (and the next one) can cause big problems if you run a lot of queries, but if we want console to have statement logging would be nice to have.

**Review by def- — COMMENTED** [review]
I marked the questionable ones with inline comments, open to keeping them at their old default. The others seem uncontroversial to me.

File: src/adapter-types/src/dyncfgs.rs
Comment (benesch) [inline src/adapter-types/src/dyncfgs.rs:24]: SGTM, but does this cause any problems for restart tests?

File: src/adapter-types/src/dyncfgs.rs
Comment (benesch) [inline src/adapter-types/src/dyncfgs.rs:31]: 👍🏽

File: src/adapter-types/src/dyncfgs.rs
Comment (benesch) [reply src/adapter-types/src/dyncfgs.rs:69]: 👍🏽 this seems very much worth doing. @ParkMyCar, want to weigh in on this though?

File: src/compute-types/src/dyncfgs.rs
Comment (benesch) [inline src/compute-types/src/dyncfgs.rs:37]: Makes sense to me but we should have @antiguru weigh in.

File: src/kafka-util/src/client.rs
Comment (benesch) [inline src/kafka-util/src/client.rs:737]: LGTM.

File: src/persist-client/src/batch.rs
Comment (benesch) [inline src/persist-client/src/batch.rs:439]: @bkirwi / @ParkMyCar can you weigh in on these?

File: src/persist-client/src/cli/admin.rs
Comment (benesch) [reply src/persist-client/src/cli/admin.rs:674]: @bkirwi can you weigh in on this one?

File: src/persist-client/src/fetch.rs
Comment (benesch) [inline src/persist-client/src/fetch.rs:1487]: @ParkMyCar

File: src/persist-client/src/internal/cache.rs
Comment (benesch) [reply src/persist-client/src/internal/cache.rs:35]: @bkirwi

File: src/persist-client/src/internal/machine.rs
Comment (benesch) [inline src/persist-client/src/internal/machine.rs:1312]: @bkirwi

File: src/persist-client/src/lib.rs
Comment (benesch) [inline src/persist-client/src/lib.rs:81]: @bkirwi

File: src/persist-client/src/stats.rs
Comment (benesch) [reply src/persist-client/src/stats.rs:32]: @bkirwi

File: src/persist/src/indexed/encoding.rs
Comment (benesch) [inline src/persist/src/indexed/encoding.rs:68]: @bkirwi / @ParkMyCar

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:470]: This all LGTM.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:783]: LGTM.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:1122]: LGTM.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:1340]: Is this @mgree? Not sure.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [reply src/sql/src/session/vars/definitions.rs:1394]: @ParkMyCar

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:1403]: @ParkMyCar

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:1773]: This is PrPr so arguably we should not enable this.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:1816]: @ggevay / @mgree

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:2015]: LGTM. Let's update the docs on this one too to indicate it's not in preview.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:2052]: @ggevay / @mgree

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:2064]: LGTM.

File: src/sql/src/session/vars/definitions.rs
Comment (benesch) [inline src/sql/src/session/vars/definitions.rs:2076]: @ggevay / @mgree

File: src/storage-types/src/dyncfgs.rs
Comment (benesch) [inline src/storage-types/src/dyncfgs.rs:211]: @petrosagg

File: src/storage-types/src/parameters.rs
Comment (benesch) [inline src/storage-types/src/parameters.rs:31]: LGTM

**Review by benesch — COMMENTED** [review]
Thanks so much for doing this, @def-! This is great. I signed off on the ones I felt comfortable signing off on, and tried to tag in the feature flag owner for all of the remaining flags.

**Comment by bkirwi** [comment]
> Is it expected?

I'm going to need to look a little harder at what that test is up to before knowing for sure... probably tomorrow!

File: src/persist-client/src/batch.rs
Comment (ParkMyCar) [inline src/persist-client/src/batch.rs:407]: Ben knows best, but I would say this should probably stay at 0 since we're not out to 256 yet in all of prod

File: src/persist-client/src/batch.rs
Comment (ParkMyCar) [reply src/persist-client/src/batch.rs:439]: `persist_inline_writes_single_max_bytes` and `persist_inline_writes_total_max_bytes` both LGTM

File: src/persist-client/src/fetch.rs
Comment (ParkMyCar) [reply src/persist-client/src/fetch.rs:1487]: LGTM!

File: src/persist-client/src/internal/cache.rs
Comment (ParkMyCar) [reply src/persist-client/src/internal/cache.rs:35]: We rolled this out to 100% a few months ago, Ben should definitely way in but I feel like this one is quite safe

File: src/persist/src/indexed/encoding.rs
Comment (ParkMyCar) [reply src/persist/src/indexed/encoding.rs:68]: Personally I would opt to keep this at `BatchColumnarFormat::Row`, but Ben is driving the rollout so will defer to him!

File: src/sql/src/session/vars/definitions.rs
Comment (ParkMyCar) [reply src/sql/src/session/vars/definitions.rs:1394]: This LGTM, if folks are having CPU issues on the `mz_catalog_server` we can suggest they reduce these statement logging related params

File: src/sql/src/session/vars/definitions.rs
Comment (ParkMyCar) [reply src/sql/src/session/vars/definitions.rs:1403]: LGTM

**Review by ParkMyCar — COMMENTED** [review]
Great to get these more aligned!

File: src/adapter-types/src/dyncfgs.rs
Comment (def-) [reply src/adapter-types/src/dyncfgs.rs:24]: We already explicitly set this to true/false in all of our tests instead of relying on the default value, so no concern.

File: src/sql/src/session/vars/definitions.rs
Comment (def-) [reply src/sql/src/session/vars/definitions.rs:1340]: @ParkMyCar I think

File: src/compute-types/src/dyncfgs.rs
Comment (antiguru) [reply src/compute-types/src/dyncfgs.rs:37]: LGTM

File: src/sql/src/session/vars/definitions.rs
Comment (ggevay) [reply src/sql/src/session/vars/definitions.rs:1816]: This is fine. I think we could even completely remove this flag. cc @bkirwi

File: src/sql/src/session/vars/definitions.rs
Comment (ggevay) [reply src/sql/src/session/vars/definitions.rs:2052]: This is fine to be on by default. (But I wouldn't completely remove it yet. There were no complaints about this so far, but it involves some tricky heuristics, so something might still come up.)

File: src/sql/src/session/vars/definitions.rs
Comment (ggevay) [reply src/sql/src/session/vars/definitions.rs:2076]: I would leave this off for self-hosted! There are various issues, e.g., https://github.com/MaterializeInc/database-issues/issues/8396

Also, this has the potential for a substantial improvement only with `enable_eager_delta_joins`, but that's not turned on, because that also has various issues.

File: src/persist-client/src/batch.rs
Comment (bkirwi) [reply src/persist-client/src/batch.rs:407]: Agreed!

File: src/persist-client/src/cli/admin.rs
Comment (bkirwi) [reply src/persist-client/src/cli/admin.rs:674]: I'll cc @jkosh44 as well, but personally I don't think it's super valuable to set this to a highish value indefinitely, especially for new envs that were never affected by the issue that we added it for. I'd personally leave this one alone...

File: src/persist-client/src/internal/cache.rs
Comment (bkirwi) [reply src/persist-client/src/internal/cache.rs:35]: Yeah, I wasn't wired in to much of the tuning here, but I think this seems reasonable to me. (It's about the size of one max-size persist part for the entire process, and if that's a hardship memory-wise we're likely not doing well on other axes.)

File: src/persist-client/src/internal/machine.rs
Comment (bkirwi) [reply src/persist-client/src/internal/machine.rs:1312]: Seems fine!

File: src/persist-client/src/lib.rs
Comment (bkirwi) [reply src/persist-client/src/lib.rs:81]: Approved!

File: src/persist-client/src/stats.rs
Comment (bkirwi) [reply src/persist-client/src/stats.rs:32]: Works for me.

File: src/persist/src/indexed/encoding.rs
Comment (bkirwi) [reply src/persist/src/indexed/encoding.rs:68]: Yeah, this isn't fully rolled out yet - let's hold off on this.

File: src/sql/src/session/vars/definitions.rs
Comment (bkirwi) [reply src/sql/src/session/vars/definitions.rs:1816]: I'm comfortable with that!

File: src/persist-client/src/cli/admin.rs
Comment (jkosh44) [reply src/persist-client/src/cli/admin.rs:674]: It might be worth setting it to a low non-zero value like 1024 though so we can ensure that the catalog shard remains compacted. Or do you think it's unlikely that the shard will become severely uncompacted again?

Comment (jkosh44) [reply src/persist-client/src/cli/admin.rs:674]: Whatever we set it to we should also update the launch darkly flag to match because in prod it's currently sitting at 131,072.

**Review by jkosh44 — APPROVED** [review]

**Comment by def-** [comment]
@bkirwi The cargo test failures still look concerning to me: [Cargo test](https://buildkite.com/materialize/test/builds/94276#01930c9e-6bd1-429a-b11d-d6dd01c8948d):
```
thread 'test_old_storage_usage_records_are_reaped_on_restart' panicked at src/environmentd/tests/server.rs:1502:9:
Records were not reaped!
thread 'test_persist_debug' panicked at src/catalog/tests/debug.rs:158:5:
assertion `left == right` failed: opening a debug catalog should not modify the contents
thread 'batch::tests::batch_builder_keys' panicked at src/persist-client/src/internal/state.rs:462:18:
expected hollow part!
thread 'batch::tests::structured_lowers' panicked at src/persist-client/src/batch.rs:1825:9:
assertion failed: part.structured_key_lower().is_some()
thread 'batch::tests::batch_builder_partial_order' panicked at src/persist-client/src/internal/state.rs:462:18:
expected hollow part!
```
Can you check if these are ok? If the tests just depend on specific config flags we can set them for the individual tests.

**Comment by def-** [comment]
@ParkMyCar Are you familiar with the statement logging? I'm not sure if this test failure is ok and we should just switch the config for the specific test:
```
thread 'test_statement_logging_immediate' panicked at src/environmentd/tests/server.rs:272:5:
assertion `left == right` failed
  left: 22
 right: 21
```

**Comment by bkirwi** [comment]
The three `batch::tests` are almost certainly just sensitive to config values, yeah.

[Reverting this change](https://github.com/MaterializeInc/materialize/pull/30385#discussion_r1833478563) may help, but otherwise I'll take a closer look at the specific tests.

**Comment by ggevay** [comment]
Random thought: Maybe this would be a good time to tune `persist_fetch_semaphore_permit_adjustment`. See discussion here: https://materializeinc.slack.com/archives/C0761MZ3QD9/p1725563934889569?thread_ts=1725561709.999339&cid=C0761MZ3QD9
(Unfortunately, I got preempted before doing the experiment on PrizePicks, and then it kind of fell through the cracks.)

File: src/sql/src/session/vars/definitions.rs
Comment (morsapaes) [reply src/sql/src/session/vars/definitions.rs:1816]: Created an issue just so we remember to remove it at some point! https://github.com/MaterializeInc/database-issues/issues/8740

File: src/compute-types/src/dyncfgs.rs
Comment (morsapaes) [reply src/compute-types/src/dyncfgs.rs:37]: @antiguru: how does this work in practice? Does `lgalloc` on Docker require attaching a volume to the container?

Comment (antiguru) [reply src/compute-types/src/dyncfgs.rs:37]: It "just works." The image has write access to the overlay, so it'll just use that for its storage. Not ideal, but we could teach users how to mount a different directory for scratch storage

**Review by def- — COMMENTED** [review]
I plan to merge this later today assuming no complaints until then. Marking one test changes inline which I'm not 100% sure about.

**Comment by ParkMyCar** [comment]
> @ParkMyCar Are you familiar with the statement logging? I'm not sure if this test failure is ok and we should just switch the config for the specific test:
> 
> ```
> thread 'test_statement_logging_immediate' panicked at src/environmentd/tests/server.rs:272:5:
> assertion `left == right` failed
>   left: 22
>  right: 21
> ```

This test has been flaky recently. I spent some time investigating it but couldn't figure out what was going on. The test relies on some random selection of data points though so the changes you made look good to me

File: src/environmentd/tests/server.rs
Comment (bkirwi) [reply src/environmentd/tests/server.rs:1460]: This setting really shouldn't change the output of anything AIUI... maybe there's a timing issue?

My understanding is that these records are not reaped synchronously, so it seems like there could be some race here, but I'm not really an expert on this code. Maybe @jkosh44?

Comment (jkosh44) [reply src/environmentd/tests/server.rs:1460]: Yeah this parameter should have no effect on records reaped. Increasing this will increase the likelihood of compacting already reaped records. We're supposed to read records only during startup. I agree with @bkirwi there's likely some other race going on.

Comment (jkosh44) [reply src/environmentd/tests/server.rs:1460]: This test is notoriously flakey and relies on background work. I would increase the timeout on line 1471 from 5s to 30s.

Comment (bkirwi) [reply src/environmentd/tests/server.rs:1460]: That sounds good to me!

Comment (def-) [reply src/environmentd/tests/server.rs:1460]: Increasing the timeout did not work! I'll open a separate issue for us to dig deeper into this when time allows.

File: src/compute-types/src/dyncfgs.rs
Comment (def-) [reply src/compute-types/src/dyncfgs.rs:37]: Since this is probably hard to use correctly, and easy to misuse and go out of disk, we'll opt to disable for now.

**Comment by def-** [comment]
I have opened two follow-up issues about the test issues I couldn't figure out:
https://github.com/MaterializeInc/database-issues/issues/8742
https://github.com/MaterializeInc/database-issues/issues/8743

**Pull request merged**
